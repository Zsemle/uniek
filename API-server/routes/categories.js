const path = require('path');

const categoriesRoutes = (app, fs) => {
  const dataPath = path.join(__dirname, '..', 'data', 'categories.json');

  const readFile = (
    callback,
    returnJson = false,
    filePath = dataPath,
    encoding = 'utf8',
  ) => {
    fs.readFile(filePath, encoding, (err, data) => {
      if (err) {
        throw err;
      }

      callback(returnJson ? JSON.parse(data) : data);
    });
  };

  app.get('/categories', (req, res) => {
    readFile((data) => {
      res.set('Access-Control-Allow-Origin', 'http://localhost:3000');
      res.send(data);
    }, true);
  });
};

module.exports = categoriesRoutes;
