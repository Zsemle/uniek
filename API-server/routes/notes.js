const path = require('path');

const notesRoutes = (app, fs) => {
  const dataPath = path.join(__dirname, '..', 'data', 'notes.json');

  const readFile = (
    callback,
    returnJson = false,
    filePath = dataPath,
    encoding = 'utf8',
  ) => {
    fs.readFile(filePath, encoding, (err, data) => {
      if (err) {
        throw err;
      }

      callback(returnJson ? JSON.parse(data) : data);
    });
  };

  app.get('/notes', (req, res) => {
    readFile((data) => {
      res.set('Access-Control-Allow-Origin', 'http://localhost:3000');
      res.send(data);
    }, true);
  });
};

module.exports = notesRoutes;
