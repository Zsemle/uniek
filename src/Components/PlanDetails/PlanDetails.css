.plan-details--subcategory-container {
  background-color: green;
}