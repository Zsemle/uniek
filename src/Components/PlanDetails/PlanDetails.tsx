import React from 'react';
import Accordion from '@mui/material/Accordion';
import AccordionSummary from '@mui/material/AccordionSummary';
import AccordionDetails from '@mui/material/AccordionDetails';
import Typography from '@mui/material/Typography';
import ExpandMoreIcon from '@mui/icons-material/ExpandMore';
import { Category, Note } from '../../types/uniekModels';
import './PlanDetails.css';

const T: { [key: string]: string } = {
  categoriesTitle: 'Your subjects:',
  categoriesEmpty: 'Your subjects will be displayed here, once you have added them to your plan.',
  notesSectionTitle: 'Notes:',
  subCategoriesSectionTitle: 'Subjects within:',
};

export interface PlanDetailsProps {
  categories: Category[] | null
  notes: Note[] | null
}

const PlanDetails:React.FC<PlanDetailsProps> = ({ categories, notes }):JSX.Element => {
  function compareCategories(a:Category, b:Category):number {
    if (a.name < b.name) {
      return -1;
    }
    if (a.name > b.name) {
      return 1;
    }
    return 0;
  }

  function compareNotes(a:Note, b:Note):number {
    if (a.dateCreated < b.dateCreated) {
      return -1;
    }
    if (a.dateCreated > b.dateCreated) {
      return 1;
    }
    return 0;
  }

  if (categories?.length === 0) {
    return (
      <div className="plan-details">
        {T.categoriesEmpty}
      </div>
    );
  }

  return (
    <div className="plan-details">
      <h2>{T.categoriesTitle}</h2>
      {categories?.sort(compareCategories)
        .filter((category:Category) => !category.parentNoteCategoryId)
        .map((category:Category) => (
          <Accordion key={`category-${category.id}`}>
            <AccordionSummary
              expandIcon={<ExpandMoreIcon />}
              aria-controls={`panel1-${category.id}-content`}
              id={`panel1-${category.id}-header`}
            >
              <Typography>{category.name}</Typography>
            </AccordionSummary>
            <AccordionDetails>
              <h4>{T.notesSectionTitle}</h4>
              <ul>
                {notes?.sort(compareNotes)
                  .filter((note:Note) => note.categoryId === category.id)
                  .map((note:Note) => <li key={`note-${note.id}`}>{note.content}</li>)}
              </ul>
              <h4>{T.subCategoriesSectionTitle}</h4>
              {categories
                .sort(compareCategories)
                .filter((subCategory:Category) => subCategory.parentNoteCategoryId === category.id)
                .map((subCategory:Category) => (
                  <Accordion
                    sx={{
                      backgroundColor: '#f1f1f1',
                    }}
                    key={`category-${subCategory.id}`}
                  >
                    <AccordionSummary
                      expandIcon={<ExpandMoreIcon />}
                      aria-controls={`panel1-${subCategory.id}-content`}
                      id={`panel1-${subCategory.id}-header`}
                    >
                      <Typography>{subCategory.name}</Typography>
                    </AccordionSummary>
                    <AccordionDetails>
                      <h4>{T.notesSectionTitle}</h4>
                      <ul>
                        {notes?.sort(compareNotes)
                          .filter((note:Note) => note.categoryId === subCategory.id)
                          .map((note:Note) => <li key={`note-${note.id}`}>{note.content}</li>)}
                      </ul>
                    </AccordionDetails>
                  </Accordion>
                ))}
            </AccordionDetails>
          </Accordion>
        ))}
    </div>
  );
};

export default PlanDetails;
