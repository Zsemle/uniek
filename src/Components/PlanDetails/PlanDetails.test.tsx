import { shallow, ShallowWrapper } from 'enzyme';
import React from 'react';
import PlanDetails, { PlanDetailsProps } from './PlanDetails';
import categories from '../../mocks/categories';
import notes from '../../mocks/notes';

describe('the plan details component', () => {
  it('renders the component with a lot of mock data', () => {
    const props:PlanDetailsProps = {
      categories,
      notes,
    };
    const wrapper:ShallowWrapper = shallow(<PlanDetails {...props} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('renders the component when the user has no categories', () => {
    const props:PlanDetailsProps = {
      categories: [],
      notes: [],
    };
    const wrapper:ShallowWrapper = shallow(<PlanDetails {...props} />);
    expect(wrapper).toMatchSnapshot();
  });
});
