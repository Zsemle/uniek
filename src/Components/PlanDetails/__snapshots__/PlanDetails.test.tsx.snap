// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`the plan details component renders the component when the user has no categories 1`] = `
<div
  className="plan-details"
>
  Your subjects will be displayed here, once you have added them to your plan.
</div>
`;

exports[`the plan details component renders the component with a lot of mock data 1`] = `
<div
  className="plan-details"
>
  <h2>
    Your subjects:
  </h2>
  <ForwardRef(Accordion)
    key="category-7"
  >
    <ForwardRef(AccordionSummary)
      aria-controls="panel1-7-content"
      expandIcon={<Memo />}
      id="panel1-7-header"
    >
      <ForwardRef(Typography)>
        Individual
      </ForwardRef(Typography)>
    </ForwardRef(AccordionSummary)>
    <ForwardRef(AccordionDetails)>
      <h4>
        Notes:
      </h4>
      <ul>
        <li
          key="note-107"
        >
          ...
        </li>
      </ul>
      <h4>
        Subjects within:
      </h4>
    </ForwardRef(AccordionDetails)>
  </ForwardRef(Accordion)>
  <ForwardRef(Accordion)
    key="category-4"
  >
    <ForwardRef(AccordionSummary)
      aria-controls="panel1-4-content"
      expandIcon={<Memo />}
      id="panel1-4-header"
    >
      <ForwardRef(Typography)>
        Language
      </ForwardRef(Typography)>
    </ForwardRef(AccordionSummary)>
    <ForwardRef(AccordionDetails)>
      <h4>
        Notes:
      </h4>
      <ul />
      <h4>
        Subjects within:
      </h4>
      <ForwardRef(Accordion)
        key="category-5"
        sx={
          Object {
            "backgroundColor": "#f1f1f1",
          }
        }
      >
        <ForwardRef(AccordionSummary)
          aria-controls="panel1-5-content"
          expandIcon={<Memo />}
          id="panel1-5-header"
        >
          <ForwardRef(Typography)>
            Reading
          </ForwardRef(Typography)>
        </ForwardRef(AccordionSummary)>
        <ForwardRef(AccordionDetails)>
          <h4>
            Notes:
          </h4>
          <ul>
            <li
              key="note-104"
            >
              Pupils should be able read N words in minute
            </li>
          </ul>
        </ForwardRef(AccordionDetails)>
      </ForwardRef(Accordion)>
      <ForwardRef(Accordion)
        key="category-6"
        sx={
          Object {
            "backgroundColor": "#f1f1f1",
          }
        }
      >
        <ForwardRef(AccordionSummary)
          aria-controls="panel1-6-content"
          expandIcon={<Memo />}
          id="panel1-6-header"
        >
          <ForwardRef(Typography)>
            Spelling
          </ForwardRef(Typography)>
        </ForwardRef(AccordionSummary)>
        <ForwardRef(AccordionDetails)>
          <h4>
            Notes:
          </h4>
          <ul>
            <li
              key="note-106"
            >
              In middle of year new handbook should be used.
            </li>
            <li
              key="note-105"
            >
              Goal for current year: no typos in simple words
            </li>
            <li
              key="note-107"
            >
              Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua.
            </li>
          </ul>
        </ForwardRef(AccordionDetails)>
      </ForwardRef(Accordion)>
    </ForwardRef(AccordionDetails)>
  </ForwardRef(Accordion)>
  <ForwardRef(Accordion)
    key="category-1"
  >
    <ForwardRef(AccordionSummary)
      aria-controls="panel1-1-content"
      expandIcon={<Memo />}
      id="panel1-1-header"
    >
      <ForwardRef(Typography)>
        Mathematics
      </ForwardRef(Typography)>
    </ForwardRef(AccordionSummary)>
    <ForwardRef(AccordionDetails)>
      <h4>
        Notes:
      </h4>
      <ul />
      <h4>
        Subjects within:
      </h4>
      <ForwardRef(Accordion)
        key="category-3"
        sx={
          Object {
            "backgroundColor": "#f1f1f1",
          }
        }
      >
        <ForwardRef(AccordionSummary)
          aria-controls="panel1-3-content"
          expandIcon={<Memo />}
          id="panel1-3-header"
        >
          <ForwardRef(Typography)>
            Division
          </ForwardRef(Typography)>
        </ForwardRef(AccordionSummary)>
        <ForwardRef(AccordionDetails)>
          <h4>
            Notes:
          </h4>
          <ul>
            <li
              key="note-102"
            >
              Pupils can do operations with fractions.
            </li>
            <li
              key="note-101"
            >
              Pupils
            </li>
          </ul>
        </ForwardRef(AccordionDetails)>
      </ForwardRef(Accordion)>
      <ForwardRef(Accordion)
        key="category-2"
        sx={
          Object {
            "backgroundColor": "#f1f1f1",
          }
        }
      >
        <ForwardRef(AccordionSummary)
          aria-controls="panel1-2-content"
          expandIcon={<Memo />}
          id="panel1-2-header"
        >
          <ForwardRef(Typography)>
            Multiplication
          </ForwardRef(Typography)>
        </ForwardRef(AccordionSummary)>
        <ForwardRef(AccordionDetails)>
          <h4>
            Notes:
          </h4>
          <ul>
            <li
              key="note-103"
            >
              Pupils should remember basic multiplication table.
            </li>
          </ul>
        </ForwardRef(AccordionDetails)>
      </ForwardRef(Accordion)>
    </ForwardRef(AccordionDetails)>
  </ForwardRef(Accordion)>
  <ForwardRef(Accordion)
    key="category-8"
  >
    <ForwardRef(AccordionSummary)
      aria-controls="panel1-8-content"
      expandIcon={<Memo />}
      id="panel1-8-header"
    >
      <ForwardRef(Typography)>
        Social
      </ForwardRef(Typography)>
    </ForwardRef(AccordionSummary)>
    <ForwardRef(AccordionDetails)>
      <h4>
        Notes:
      </h4>
      <ul />
      <h4>
        Subjects within:
      </h4>
    </ForwardRef(AccordionDetails)>
  </ForwardRef(Accordion)>
</div>
`;
